proposals/NNNN-sdl-protocol-security-specification.md

```diff
@@ -0,0 +1,394 @@
+
+## Proposed solution
+
+The solution is to provide a new section into the protocol spec around security and protection. The baseline for this section is reverse engineered from SDL Core and the currently implemented behavior of the security manager but also existing documentation from SDL Core and the libraries. 
+
+### Change 1: Update Frame Info fields
+
+Because TLS handshake uses a Frame Info field for Single frames the spec must be changed accordingly
+
+<table width="100%">
+  <tr>
+    <th>Field</th>
+    <th>Size</th>
+    <th>Description</th>
+  </tr>
+  <tr>
+    <td>Frame Info</td>
+    <td>8 bit</td>
+    <td>
+      ...<br>
+      <b>Frame Type = 0x01 (Single Frame)</b><br>
+      <s>0x00 - 0xFF Reserved</s><br>
+      0x00 Single Frame<br>
+      0x01 - 0xFF Reserved<br>
+      ...
+    </td>
+<table>
+
+### Change 2: Add New Section
+
+Following section should be added to the protocol spec:
+
+## 4.7 Secured Communication
+
+It is possible to establish a secured and encrypted communication with the system by setting the frame header encryption flag to `1` when starting a new service. If the authentication was successful, the system will reply with a `StartService ACK` frame with the encryption flag also set to `1` indicating that encrypted data is now accepted. If the authentication fails for some reason the system will reset the TLS connection and return a `StartService NAK` frame.
+
+### 4.7.1 Authentication
+
+Below diagram shows the sequence of how the TLS handshake exchanges certificates to compute the master secret.
+
+![TLS Handshake activity diagram](../assets/proposals/NNNN-sdl-protocol-security-specification/tls-handshake.png)
+
+The authentication is done using TLS handshake. The TLS handshake process is defined by TLS and is not part of the SDL protocol. The handshake is designed as a client server communication which is configurable in the system settings. An application can take the role of a server where the system is the client or vice versa. This setting is not dynamic which means an SDL integrator must agree on one setup and avoid Client/Client or Server/Server connections. The client entity will initiate a TLS handshake with the corresponding security manager of the server. The client will do this only if the server was not authenticated before in the current transport connection. According to the TLS handshake process the peer certificate can be omitted for the server but it's required for the client.
+
+The system can be configured to support one encryption method. Following methods are supported:
+
+- SSLv3
+- TLSv1
+- TLSv1.1
+- TLSv1.2
+- DTLSv1
+
+Dependent of the role the system has to initiate with the corresponding server method or client method.
```

Comment:
I'm having trouble understanding this sentence.  Can you please re-word?

Reply:
This is very technical and related to the design of TLS. Roles are explained in the previous paragraph and that the system (Core) can be a server or client role. Let me try to add an example.

Reply:
```suggestion
Dependent of the role the system has to initiate with the corresponding server method or client method. For instance if the system is configured to use DTLSv1 and its role is set to client, the system has to use the method DTLSv1_client. The application role has to be server and must use DTLSv1_server.
```

Reply:
@theresalech Please check out my example.

Reply:
Thank you!  This helps to clarify.  I'd suggest a tweak to the first sentence, so committing the below:

```suggestion
Depending on the role, the system has to initiate with the corresponding server method or client method.  For instance, if the system is configured to use DTLSv1 and its role is set to client, the system has to use the method DTLSv1_client. The application role has to be server and must use DTLSv1_server.
```